ARCs/arc-0039.md

```diff
@@ -0,0 +1,54 @@
+created: 2023-07-15
+requires: 4
+---
+
+## Abstract
+According to [ARC-0004](./arc-0004.md), when encoding dynamic types in a tuple the tuple consists of a `head` containing all of the offsets of the actual values in the `tail`. An additional value, the byte offset of the end of the tuple, should be added to the end of the `head` value for more efficient element reading.
+
+## Motivation
+When reading a value in a dynamic tuple, the head offset of the element must be extracted as a `uint16` to get the start of the element. To get the end of the element, which is necessary to properly read the element, the following head value can be extracted. This, however, fails to work when the element being read is the last element in its respective dynamic array or tuple. When the element is the last in the array or tuple, one must compute the length of the element. For static types, this is trivial and for dynamic arrays of static types this is also trivial because the length prefix can be used. For more complex nested dynamic types, however, this could involve many levels of extracting lengths and offsets, thus reading an element is `O(N)`, which `N` being the depth of the nested dynamic types. If there was an additional head value containing the end of the dynamic tuple, readnig any element, regardless of its type or position, would be `O(1)` since one can always just extract the following head value.
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+Tuples **MUST** be encoded with the following rules:
+
+* If `x` is a tuple of `N` types, `(T1,T2,...,TN)`, where `x[i]` is the value at index `i`, starting at 1 and `x[i]` is dynamic:
+    * `head(x[i]) = enc(2 + len( head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[i-1]) ))`
+    * `tail(x[i]) = enc(x[i])`
+* If `x` is a tuple of `N` types, `(T1,T2,...,TN)`, and any type is dynamic
+  * `x = enc( head(x[i]) ... head(x[N]) + ( 2 + len( head(x[i]) ... head(x[N]) ) + len( tail(x[i]) ... tail(x[N]) ) ) + tail(x[i]) ... tail(x[N]) )`
+
+
+## Rationale
+This will make reading values from an ABI encoded tuple much more efficient, which is especially important for on-chain logic.
+
+## Backwards Compatibility
+[ARC-0004](./arc-0004.md) clients will be able to read tuples encoded according to this ARC, but they will not be able to write them.
+
+## Test Cases
+N/A
+
+## Reference Implementation
+
+Encoding of `["Hello", "World"]` as `string[]`
+
+[ARC-0004](./arc-0004.md): `0x0002 0004 000b 0005 48656c6c6f 0005 576f726c64`
+
+This ARC: `0x0002 0006 000d 0014 0005 48656c6c6f 0005 576f726c64`
+
+Note the addition of the `0014` signifying the end of the tuple/array and `head(x[1])` and `head[x[2]]` being incremented by two to account for this additional value.
```

Comment:
If I understand correctly, this is not a great example because `0014` points to the end of the encoding, and that's not very useful here because it's just the length of the byte string.

Can you provide an example where the benefit of including this additional information is clearer?

Reply:
Yeah this example was chosen to demonstrate the encoding rules, but it doesn't really show the rationale. Take a slightly more complex example:

`[["Hello", "World"], [1, 2, 3]]` encoded as `(string[],uint8[])`

`0004 0018 0002 0004 000b 0005 48656c6c6f 0005 576f726c64 0003 01 02 03`

To get the end of `x[0][1]` you will always need to check if `x[0][1]` is the last element of `x[0]`. If it's not, you just extract the next head but if it is you need to go up a level and get the head for `x[1]`. 

In this example, it exists but if there was further levels of nesting you would need to check if the parent element is the last element in it's parent and repeat until the next head is found. This means you need to spend opcodes checking the accessor against the prefix for every dynamic parent until you find the first parent tuple that has a dynamic element proceeding the accessed element.

In the context of TEAL compilers, the compiler has no way of knowing whether the accessed element is last or not without using up those opcodes, unless only literal array access is supported.

Reply:
Alternatively, you use the length prefix to calculate the length of the accessed element, but if that element is a nested dynamic type then you run into similar complexity of needing to calculate the total length of the last element each level.

Reply:
Why are you trying to get to the end of `x[0][1]`?

If your goal is to start decoding `x[1]`, you already know where to start, because the head of `x[1]` is in a well defined spot (bytes 2-3 of the input), and that tells you that `x[1]` starts at by 0x18.

If you are not trying to reach `x[1]` I suspect you are trying to `substr` out `x[0]` in order to give it to some decoding routine, so you'd like to know where it ends. But you don't have to do that. Just give that routine the entire suffix of the input, starting where `head(x[0])` tells you to begin (byte 4).

You say it's hard "To get the end of `x[0][1]`" and I am saying that is not an operation you should need directly. If you are going down the dynamic path of `x[0]`, you will find out where `x[0][1]` ends when you examine the head of `x[0][1]`, combined with the length where `head(x[0][1])` points. If you don't actually care about `x[0][1]` itself and are really just trying to get to the end of `x[0]`, you know where `x[1]` begins.

Reply:
The goal is to get the value of `x[0][1]` as efficiently as possible, so get the start of `x[0][1]`, the end of `x[0][1]` and use `substring3`. 

> combined with the length where head(x[0][1])

Which is great when `x[0][1]` is of type `T[]` and `T` is static, but if `T` is dynamic then you need spend opcode budget to calculate the total length of `x[0][1]`. For every level of dynamic nested types in `T`, it consumes opcodes to calculate the total length. Having a value that points to the end of `x[0][1]` in the encoded data structure makes it trivial to get the end, rather than potentially having to calculate all of the lengths.

Reply:
I explained why substr'ing out the value is not necessary.

> I suspect you are trying to substr out x[0] in order to give it to some decoding routine, so you'd like to know where it ends. But you don't have to do that. Just give that routine the entire suffix of the input, starting where head(x[0]) tells you to begin (byte 4).